File: app/src/main/java/pl/mosenko/sunnypodlaskie/util/WeatherNotificationUtil.java
package pl.mosenko.sunnypodlaskie.util;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.content.ContextCompat;

import java.text.SimpleDateFormat;
import java.util.Date;

import pl.mosenko.sunnypodlaskie.mvp.weatherdatalist.MainActivity;
import pl.mosenko.sunnypodlaskie.R;

/**
 * Created by syk on 26.05.17.
 */

public class WeatherNotificationUtil {
    private static final int WEATHER_NOTIFICATION_ID = 3011;

    public static void notifyUserOfNewWeatherData(@NonNull Context context, Date receivingTime) {
        Notification notification = buildNotification(context, receivingTime);
        notify(context, notification);
    }

    private static void notify(@NonNull Context context, Notification notification) {
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.notify(WEATHER_NOTIFICATION_ID, notification);
    }

    @NonNull
    private static Notification buildNotification(@NonNull Context context, Date receivingTime) {
        String notificationTitle = context.getString(R.string.notification_title);
        String notificationContentText = context.getString(R.string.notification_content_text);
        String receivingTimeFormatted = getFormattedReceivingDate(receivingTime);
        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(context)
                .setColor(ContextCompat.getColor(context, R.color.colorPrimary))
                .setContentTitle(notificationTitle)
                .setContentText(notificationContentText + " " + receivingTimeFormatted)
                .setSmallIcon(R.drawable.ic_clear) //app's icon
                .setAutoCancel(true);

        PendingIntent mainAcitivityIntentPendingIntent = getPendingIntent(context);
        notificationBuilder.setContentIntent(mainAcitivityIntentPendingIntent);
        return notificationBuilder.build();
    }

    private static PendingIntent getPendingIntent(@NonNull Context context) {
        Intent mainAcitivityIntent = new Intent(context, MainActivity.class);
        TaskStackBuilder taskStackBuilder = TaskStackBuilder.create(context);
        taskStackBuilder.addNextIntent(mainAcitivityIntent);
        return taskStackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
    }

    private static String getFormattedReceivingDate(Date receivingTime) {
        SimpleDateFormat notificationDateFormat = new SimpleDateFormat("MMM, dd HH:mm:ss");
        return notificationDateFormat.format(receivingTime);
    }


}

File: app/src/main/java/pl/mosenko/sunnypodlaskie/util/WeatherNotificationUtil.kt
package pl.mosenko.sunnypodlaskie.util

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import pl.mosenko.sunnypodlaskie.R
import pl.mosenko.sunnypodlaskie.mvp.MainActivity
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by syk on 26.05.17.
 */

private const val WEATHER_NOTIFICATION_ID = 3011
private const val NOTIFICATION_CHANNEL_ID = "4011"

class WeatherNotificationUtil(private val context: Context) {

    fun notifyUserOfNewWeatherData(receivingTime: Date) {
        createNotificationChannel(context)
        val notification = buildNotification(receivingTime)
        notify(notification)
    }

    private fun notify(notification: Notification) {
        val notificationManager = NotificationManagerCompat.from(context)
        notificationManager.notify(WEATHER_NOTIFICATION_ID, notification)
    }

    private fun createNotificationChannel(context: Context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = context.getString(R.string.notification_channel_name)
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, name, importance)
            val notificationManager =
                context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun buildNotification(receivingTime: Date): Notification {
        val notificationTitle = context.getString(R.string.notification_title)
        val notificationContentText = context.getString(R.string.notification_content_text)
        val receivingTimeFormatted = getFormattedReceivingDate(receivingTime)
        val notificationBuilder = NotificationCompat.Builder(context, NOTIFICATION_CHANNEL_ID)
            .setColor(ContextCompat.getColor(context, R.color.colorPrimary))
            .setContentTitle(notificationTitle)
            .setContentText("$notificationContentText $receivingTimeFormatted")
            .setSmallIcon(R.drawable.ic_clear)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)
        val mainActivityIntentPendingIntent = getPendingIntent()
        notificationBuilder.setContentIntent(mainActivityIntentPendingIntent)
        return notificationBuilder.build()
    }

    private fun getPendingIntent(): PendingIntent? {
        val mainActivityIntent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        }
        return PendingIntent.getActivity(context, 0, mainActivityIntent, 0)
    }

    private fun getFormattedReceivingDate(receivingTime: Date): String? {
        val notificationDateFormat = SimpleDateFormat("HH:mm:ss dd.MM", Locale.getDefault())
        return notificationDateFormat.format(receivingTime)
    }
}
